libclc/ptx-nvidiacl/libspirv/math/fmax.cl

```diff
@@ -6,12 +6,66 @@
 //
 //===----------------------------------------------------------------------===//
 
+#include <clcmacro.h>
 #include <spirv/spirv.h>
 
-#include "../../include/libdevice.h"
-#include <clcmacro.h>
+extern int __clc_nvvm_reflect_arch();
+
+_CLC_DEF _CLC_OVERLOAD float __spirv_ocl_fmax(float x, float y) {
+  return __nvvm_fmax_f(x, y);
+}
+_CLC_BINARY_VECTORIZE(_CLC_OVERLOAD _CLC_DEF, float, __spirv_ocl_fmax, float,
+                      float)
+
+#ifdef cl_khr_fp64
+
+#pragma OPENCL EXTENSION cl_khr_fp64 : enable
+
+_CLC_DEF _CLC_OVERLOAD double __spirv_ocl_fmax(double x, double y) {
+  return __nvvm_fmax_d(x, y);
+}
+_CLC_BINARY_VECTORIZE(_CLC_OVERLOAD _CLC_DEF, double, __spirv_ocl_fmax, double,
+                      double)
+
+#endif
+
+#ifdef cl_khr_fp16
+
+#pragma OPENCL EXTENSION cl_khr_fp16 : enable
+
+_CLC_DEF _CLC_OVERLOAD half __spirv_ocl_fmax(half x, half y) {
+  if (__clc_nvvm_reflect_arch() >= 800) {
+    return __nvvm_fmax_f16(x, y);
+  }
+  return x > y ? x : y;
```

Comment:
It doesn't seem to be correct implementation of fmax instruction.
From https://www.khronos.org/registry/SPIR-V/specs/unified1/OpenCL.ExtendedInstructionSet.100.html

> If one operand is a NaN, fmax returns the other argument.

Which means if `y` is `NaN`, it must return `x`. According to my understanding expression `x > y` returns false, but expected return value is `x`, not `y`. Right?

The same comment applies to `fmin` implementation.

Reply:
Right. I changed these back to using float implementation from libdevice.